oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/celliterator/CellIterator.java

```diff
@@ -802,6 +804,22 @@ public IterateAllEntry(
     });
   }
 
-}
-
-
+  private static SortedMap<OSHDBTimestamp, OSMEntity> getVersionsByTimestamps(
+      OSHEntity osh, List<OSHDBTimestamp> byTimestamps) {
+    SortedMap<OSHDBTimestamp, OSMEntity> result = new TreeMap<>();
+
+    int i = byTimestamps.size() - 1;
+    Iterator<? extends OSMEntity> itr = osh.getVersions().iterator();
+    while (itr.hasNext() && i >= 0) {
+      OSMEntity osm = itr.next();
+      if (osm.getTimestamp().getRawUnixTimestamp() <= byTimestamps.get(i).getRawUnixTimestamp()) {
+        while (i >= 0 && osm.getTimestamp().getRawUnixTimestamp() <= byTimestamps.get(i)
+            .getRawUnixTimestamp()) {
+          result.put(byTimestamps.get(i), osm);
+          i--;
+        }
+      }
```

Comment:
I think the first if check is unnecessary, we can reduce the nesting by one level. And for the timestamp comparison, why not use the compareTo method of the OSHDBTimestamp directly.
Here my suggestion:
```java
while (itr.hasNext() && i >= 0) {
  OSMEntity osm = itr.next();
  while (i >= 0 && osm.getTimestamp().compareTo(byTimestamps.get(i)) <= 0) {
    result.put(byTimestamps.get(i), osm);
    i--;
  }
}
```

Reply:
:thinking: We have to check this thoroughly. At least the unit tests manage to hit both cases of the `if` statement:

![image](https://user-images.githubusercontent.com/1927298/107407110-8aa6d880-6b09-11eb-96d6-7ee5852f3d5c.png)

Reply:
ah, wait. I see what you mean now :see_no_evil:  The `while` condition is equivalent to the surrounding `if`. Sure, then it can certainly be omitted.

Reply:
could we omitted that redundant if check, and use compareTo from the OSHDBTimestamp instead checking against the rawUnixTimestamp?

Reply:
yes, see https://github.com/GIScience/oshdb/pull/332/commits/5a26b21633d3df7785e98b7aa1e1737bf3fcc5c1#diff-fbb13d5279408199848215dfdc8b65baafc3725c8a64c579d9ca119aeba12cc6R819